tests/schema_test.py

```diff
@@ -6,37 +6,60 @@
 schema.
 
 """
+import collections
+import unittest
 
 from .utils import BaseTest, check_for_cycles, validate_schemata
 
 
 class SchemaTest(BaseTest):
 
-    def traverse_schema_with_conditional(self, root, conditional_func, test_func):
-        if conditional_func(root):
-            test_func(root)
+    @unittest.expectedFailure
+    def test_enum(self):
+        stack = collections.deque()
+        res = collections.defaultdict(list)
 
-        if isinstance(root, list):
-            for child in root:
-                self.traverse_schema_with_conditional(child, conditional_func, test_func)
-        elif isinstance(root, dict):
-            for child in root.keys():
-                self.traverse_schema_with_conditional(root[child], conditional_func, test_func)
+        def _get_path_str(path_stack):
+            return '->'.join(path_stack)
 
-    def test_enum_is_string(self):
-
-        def trigger(root):
+        def bfs_with_enum_test(root, test_func, path):
```

Comment:
Couldn't we make this a static class method? To me these functions encapsulated in the test function add a lot of clutter and make it hard to read. Also would you be able to give them a doc string? bfs_with_enums just reads as best friends with enums and not exactly sure what this is doing lol :P 

Also this approach relies completely on side effect with all of these methods building on and appending common objects. I'd challenge you to take a more functional approach build things up and have meaningful returns with no side effect outside of their own scope.

Reply:
I don't think bfs is the right terminology for going through the node definitions in `self.dictionary.schema`. It doesn't have to be recursive too, I am guessing someone just wanted to be fancy when they did the old implementation. Some other way to look at this would be something like this
``` python
for node, node_dict in self.dictionary.schema.items():
    for prop, prop_dict in node_dict["properties"].items():
        if "enum" in prop_dict:
            # do some magic
```